docs/tips-and-tricks/debugging-toolkit.md

```diff
@@ -0,0 +1,191 @@
+-- Experiment: Try different time ranges to see what suspicious data patterns emerge
+SELECT 
+    'Data Quality Check' as analysis,
+    data_year,
+    count() as events,
+    CASE 
+        WHEN data_year < 2010 THEN 'BAD: Suspiciously old timestamps'
+        WHEN data_year > 2030 THEN 'BAD: Far future timestamps'  
+        ELSE 'NORMAL'
+    END as data_quality
+FROM (
+    SELECT toYear(created_at) as data_year
+    FROM github.github_events 
+    WHERE created_at >= '2020-01-01'
+)
+GROUP BY data_year
+ORDER BY data_year DESC;
+```
+
+## The 2AM Methodology {#the-2am-methodology}
+
+**Follow this exact sequence when everything is broken:**
+
+### Phase 1: Immediate Triage (30 seconds) {#phase-1-immediate-triage}
+
+1. Run `system.errors` - any non-zero = active incident
+2. Check disk space - *"It took us from 12 to 4 AM... it was as simple as low disk"*
+3. Look for replication lag > 5 minutes
+
+### Phase 2: Resource Investigation (2 minutes) {#phase-2-resource-investigation}
+
+4. Find memory-hungry queries in `system.processes`
+5. Check for stuck merges running >1 hour
+6. Kill obviously problematic queries
+
+### Phase 3: Data Quality Check (5 minutes) {#phase-3-data-quality-check}
+
+7. Look for bad partitions (1998, 2050 dates)
+8. Check for parts explosion (>1000 parts per table)
+
+## Emergency Actions Reference {#emergency-actions}
+
+**Production-tested solutions:**
+
+| Problem | Detection Query | Solution |
+|---------|-----------------|----------|
+| **Memory OOM** | `SELECT * FROM system.processes WHERE memory_usage > 8GB` | Enable `external_aggregation=1` |
+| **Disk Full** | `SELECT sum(bytes_on_disk) FROM system.parts` | Delete old partitions, expand disk |
+| **Replication Lag** | `SELECT * FROM system.replicas WHERE absolute_delay > 300` | Check network, restart lagging replica |
+| **Stuck Query** | `SELECT * FROM system.processes WHERE elapsed > 300` | `KILL QUERY WHERE query_id = '...'` |
+| **Parts Explosion** | `SELECT count() FROM system.parts WHERE active=1` | Enable async_insert, increase batch sizes |
+
+**The golden rule:** *"Problems very rarely just pop out of nowhere there are signs... investigate before it goes from 15 milliseconds to 30 seconds"*
+
+## Community War Stories & Lessons {#war-stories}
+
+**Disk Space Issues:**
+- *"Default AWS EBS limit of 16TB hits you when you least expect it"*
+- *"Other nodes keep sending data to the full node creating cascading failure"*
+
+**Memory Exhaustion:**  
+- *"Out of memory typically appears when you have a big aggregation with a lot of keys"*
+- *"Enable external aggregation - query will be slower but won't crash"*
+
+**Bad Data:**
+- *"Developers send data from 1998 or 2050 causing partition chaos"*
+- *"Always validate timestamps before they hit production"*
+
+**The key insight:** Most 2AM incidents are preventable if you recognize the warning signs and have ready-to-use diagnostic queries.
```

Comment:
This feels hidden a little (applies to other pages). Maybe we put these key insights in an admonition which will render big and green:

```suggestion
:::tip The key insight
Most 2AM incidents are preventable if you recognize the warning signs and have ready-to-use diagnostic queries.
:::
```